<a href="https://colab.research.google.com/github/walterlopez32/Laboratorio-Analisis-de-Dato/blob/main/Laboratorio1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Laboratorio 1


#ETL (Extraccion, Transformacion y Carga)

## 1. Extraccion( Extract) sintáxis para montar y hacer la extraccion



In [83]:
from google.colab import files
import pandas as pd

# Abrir el explorador de archivos y seleccionar el archivo
uploaded = files.upload()




In [51]:
from google.colab import drive
import pandas as pd
import chardet

# Montar Google Drive
drive.mount('/content/drive', force_remount=True)

# Rutas del archivos CSV
ruta_archivo = "/content/drive/MyDrive/Finance_Trends_enriquecido.csv"


# Función para detectar la codificación
def detectar_codificacion(ruta):
    with open(ruta, 'rb') as f:
        result = chardet.detect(f.read())
    encoding = result['encoding']
    print(f" {ruta.split('/')[-1]} → Codificación detectada: {encoding}")
    return encoding

# Función para leer CSV con detección automática
def leer_csv(ruta):
    try:
        encoding = detectar_codificacion(ruta)

        df = pd.read_csv(ruta, sep= None, engine= "python", encoding=encoding)
        # sep=None con engine='python' → pandas detecta automáticamente si el separador es , o ;.
        print(f" {ruta.split('/')[-1]} cargado correctamente. ({df.shape[0]} filas, {df.shape[1]} columnas)")
        return df
    except FileNotFoundError:
        print(f" Error: No se encontró el archivo en {ruta}")
        return None
    except Exception as e:
        print(f" Error inesperado al leer {ruta.split('/')[-1]}: {e}")
        return None

#  Leer archivos
df_Finanzas = leer_csv(ruta_archivo)




Mounted at /content/drive
 Finance_Trends_enriquecido.csv → Codificación detectada: utf-8
 Finance_Trends_enriquecido.csv cargado correctamente. (12000 filas, 32 columnas)


## 2. Transformación de Datos

In [53]:
#df_comercio.head()
df_Finanzas.head(10)
#df_sucursales.head()

gender  age Investment_Avenues  Mutual_Funds  Equity_Market  Debentures  \
0    Male   29                Yes             2              4           7   
1    Male   28                Yes             2              3           6   
2  Female   28                Yes             2              3           7   
3    Male   19                Yes             2              3           7   
4    Male   32                Yes             2              4           7   
5    Male   31                Yes             2              4           7   
6    Male   27                Yes             3              2           7   
7    Male   30                Yes             3              4           6   
8  Female   37                Yes             2              4           7   
9  Female   37                Yes             2              4           7   

   Government_Bonds  Fixed_Deposits  PPF  Gold  ...      Reason_FD  \
0                 5               3    1     6  ...  Fixed Returns   
1                 5               1    4     7  ...      Risk Free   
2                 5               4    1     6  ...      Risk Free   
3                 4               6    1     5  ...      Risk Free   
4                 5               3    1     6  ...  Fixed Returns   
5                 5               3    1     6  ...  Fixed Returns   
6                 4               5    1     6  ...      Risk Free   
7                 5               2    1     7  ...      Risk Free   
8                 5               3    1     6  ...      Risk Free   
9                 5               3    1     6  ...      Risk Free   

                     Source      País      Banco Tipo_de_activo Moneda  \
0  Newspapers and Magazines    Brasil       Itaú         Fondos    USD   
1                Television     Chile       BBVA          Bonos    EUR   
2     Financial Consultants    México       HSBC  Criptomonedas    ARS   
3  Newspapers and Magazines    España  Santander  Criptomonedas    ARS   
4     Financial Consultants    Brasil       HSBC       Acciones    BRL   
5  Newspapers and Magazines  Colombia     Nación       Acciones    BRL   
6  Newspapers and Magazines      EEUU    Galicia          Bonos    BRL   
7     Financial Consultants    México      Macro  Criptomonedas    EUR   
8     Financial Consultants     Chile       HSBC    Commodities    EUR   
9     Financial Consultants    México       Itaú          Bonos    BRL   

  Tasa_interes Riesgo                          Fecha ID_Registro  
0         2.90   Bajo  2022-01-01 00:00:00.000000000           1  
1         3.95   Alto  2022-01-01 02:48:00.840070005           2  
2         1.71   Alto  2022-01-01 05:36:01.680140011           3  
3         5.39   Bajo  2022-01-01 08:24:02.520210017           4  
4         2.37   Alto  2022-01-01 11:12:03.360280023           5  
5         6.08  Medio  2022-01-01 14:00:04.200350029           6  
6         2.82  Medio  2022-01-01 16:48:05.040420035           7  
7         3.33   Bajo  2022-01-01 19:36:05.880490040           8  
8        11.98   Bajo  2022-01-01 22:24:06.720560046           9  
9         0.88   Bajo  2022-01-02 01:12:07.560630052          10  

[10 rows x 32 columns]

In [54]:
# Hcemos un .info al Data Frame para ver las cantidad de registros y columnas,  si hay valores nulos
# y los tipos de datos para ver si nesecitamos cambiaar alguno en especifico
# Realizando esta sintaxis también vemos que los nombres de las columnas estas en otro idioma
df_Finanzas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12000 entries, 0 to 11999
Data columns (total 32 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   gender                             12000 non-null  object 
 1   age                                12000 non-null  int64  
 2   Investment_Avenues                 12000 non-null  object 
 3   Mutual_Funds                       12000 non-null  int64  
 4   Equity_Market                      12000 non-null  int64  
 5   Debentures                         12000 non-null  int64  
 6   Government_Bonds                   12000 non-null  int64  
 7   Fixed_Deposits                     12000 non-null  int64  
 8   PPF                                12000 non-null  int64  
 9   Gold                               12000 non-null  int64  
 10  Stock_Marktet                      12000 non-null  object 
 11  Factor                             12000 non-null  obj

In [55]:
# Detectamos que en una de las columnas hay un error ortográfico por lo cual pasamos a corregir
# Cargar el dataset
df_Finanzas = pd.read_csv("/content/drive/MyDrive/Finance_Trends_enriquecido.csv")

# Corregir error de nombre
df_Finanzas = df_Finanzas.rename(columns={"Stock_Marktet": "Stock_Market"})

print(" La correccón se realizó correctamente")
display(df_Finanzas[["Stock_Market"]].head())
#df_Cambio_Idioma.head()





 La correccón se realizó correctamente


Stock_Market
0          Yes
1          Yes
2          Yes
3          Yes
4          Yes

In [56]:
df_Finanzas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12000 entries, 0 to 11999
Data columns (total 32 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   gender                             12000 non-null  object 
 1   age                                12000 non-null  int64  
 2   Investment_Avenues                 12000 non-null  object 
 3   Mutual_Funds                       12000 non-null  int64  
 4   Equity_Market                      12000 non-null  int64  
 5   Debentures                         12000 non-null  int64  
 6   Government_Bonds                   12000 non-null  int64  
 7   Fixed_Deposits                     12000 non-null  int64  
 8   PPF                                12000 non-null  int64  
 9   Gold                               12000 non-null  int64  
 10  Stock_Market                       12000 non-null  object 
 11  Factor                             12000 non-null  obj

In [57]:
# Detectamos que los nombres de las  columnas estan en otro idioma por la cual procedemos a cambiar al español
# Con esta sintaxis cambiamos el idioma de cada columna para
# tener un mejor entendimineto de los conceptos y las columnas

#Cargar el dataset
df_Cambio_Idioma = df_Finanzas.copy()


# Diccionario de traducción de columnas
traduccion = {
    "gender": "Genero",
    "age": "Edad",
    "Investment_Avenues": "Vias_de_inversion",
    "Mutual_Funds": "Fondos_mutuos",
    "Equity_Market": "Mercado_de_valores",
    "Debentures": "Obligaciones - Bonos Corporativos",
    "Government_Bonds": "Bonos_gubernamentales",
    "Fixed_Deposits": "Depositos_Plazos_fijos",
    "PPF": "Fondo_de_prevision",
    "Gold": "Inversión_Oro",
    "Stock_Market": "Part_En_Mercado_de_acciones",
    "Factor": "Factor que influye en Invertir ",
    "Objective": "Objetivo",
    "Purpose": "Proposito",
    "Duration": "Duracion",
    "Invest_Monitor": "Frecuencia_monitoreo",
    "Expect": "Expectativa_rendimiento",
    "Avenue": "Tipo_inversion(Corto o largo Plazo,etc)",
    "What are your savings objectives?": "Objetivos_ahorro",
    "Reason_Equity": "Razon_equidad",
    "Reason_Mutual": "Razon_fondos_mutuos",
    "Reason_Bonds": "Razon_bonos",
    "Reason_FD": "Razon_depositos_fijos",
    "Source": "Fuente_informacion"
}

# Aplicar traducción
df_Cambio_Idioma = df_Cambio_Idioma.rename(columns=traduccion)

# Normalizar texto (opcional)

for col in df_Cambio_Idioma.columns:
    if df_Cambio_Idioma[col].dtype == 'object':
        df_Cambio_Idioma[col] = df_Cambio_Idioma[col].map(lambda x: x.strip().capitalize() if isinstance(x, str) else x)


# Ruta de salida para guardar el archivo limpio y traducido
ruta_salida = "/content/drive/MyDrive/Finance_Trends_enriquecido_limpio_traducido.csv"

# Guardar el DataFrame limpio y traducido como CSV
df_Cambio_Idioma.to_csv(ruta_salida, index=False, encoding="utf-8-sig")


print(f" Archivo limpio, traducido y guardado correctamente en '{ruta_salida}'")

 Archivo limpio, traducido y guardado correctamente en '/content/drive/MyDrive/Finance_Trends_enriquecido_limpio_traducido.csv'


In [58]:
# Seleccionar columnas de tipo 'object'
# para lugo poder cambiar el tipo de dato
object_columns = df_Cambio_Idioma.select_dtypes(include='object').columns

# Mostrar valores únicos para las columnas object (se puede no mostrar demasiados si hay muchas)
for col in object_columns[:32]:  #  permite limitar las columnas si queremos
    print(f"Columna: {col}")
    print(df_Cambio_Idioma[col].unique())
    print("-" * 30) # permite realizar el saparador entre columans y sus registros


Columna: Genero
['Male' 'Female']
------------------------------
Columna: Vias_de_inversion
['Yes' 'No']
------------------------------
Columna: Part_En_Mercado_de_acciones
['Yes' 'No']
------------------------------
Columna: Factor que influye en Invertir 
['Risk' 'Returns' 'Locking period']
------------------------------
Columna: Objetivo
['Growth' 'Capital appreciation' 'Income']
------------------------------
Columna: Proposito
['Wealth creation' 'Savings for future' 'Returns']
------------------------------
Columna: Duracion
['Less than 1 year' '3-5 years' 'More than 5 years' '1-3 years']
------------------------------
Columna: Frecuencia_monitoreo
['Weekly' 'Daily' 'Monthly']
------------------------------
Columna: Expectativa_rendimiento
['20%-30%' '30%-40%' '10%-20%']
------------------------------
Columna: Tipo_inversion(Corto o largo Plazo,etc)
['Public provident fund' 'Mutual fund' 'Equity' 'Fixed deposits']
------------------------------
Columna: Objetivos_ahorro
['Health c

In [59]:

# Cambiar el tipo de dato de la columna 'Fecha' a 'datetime' ya que aparece como abject
df_Cambio_Idioma['Fecha'] = pd.to_datetime(df_Cambio_Idioma['Fecha'])

# Verificar los tipos de datos actualizados
print("Tipos de datos después de la conversión:")
print(df_Cambio_Idioma[['Fecha']].info())

Tipos de datos después de la conversión:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12000 entries, 0 to 11999
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   Fecha   12000 non-null  datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 93.9 KB
None


In [60]:
# Convertir las columnas que estaban tipo object al tipo 'category'
# ya que en una df de muchos registro permite utilizar menos recursos cuando hay datos repetidos
# Seleccionar columnas de tipo 'object'
object_columns = df_Cambio_Idioma.select_dtypes(include='object').columns

for col in object_columns:
    if col in df_Cambio_Idioma.columns: # Verificar si la columna existe en el DataFrame
        df_Cambio_Idioma[col] = df_Cambio_Idioma[col].astype('category')
        print(f"Columna '{col}' convertida a tipo 'category'.")
    else:
        print(f"Advertencia: La columna '{col}' no se encontró en el DataFrame.")

print("\n Conversión a tipo 'category' completada.")

# Mostrar los tipos de datos actualizados para verificar
print("\nTipos de datos después de la conversión:")
print(df_Cambio_Idioma.info())

Columna 'Genero' convertida a tipo 'category'.
Columna 'Vias_de_inversion' convertida a tipo 'category'.
Columna 'Part_En_Mercado_de_acciones' convertida a tipo 'category'.
Columna 'Factor que influye en Invertir ' convertida a tipo 'category'.
Columna 'Objetivo' convertida a tipo 'category'.
Columna 'Proposito' convertida a tipo 'category'.
Columna 'Duracion' convertida a tipo 'category'.
Columna 'Frecuencia_monitoreo' convertida a tipo 'category'.
Columna 'Expectativa_rendimiento' convertida a tipo 'category'.
Columna 'Tipo_inversion(Corto o largo Plazo,etc)' convertida a tipo 'category'.
Columna 'Objetivos_ahorro' convertida a tipo 'category'.
Columna 'Razon_equidad' convertida a tipo 'category'.
Columna 'Razon_fondos_mutuos' convertida a tipo 'category'.
Columna 'Razon_bonos' convertida a tipo 'category'.
Columna 'Razon_depositos_fijos' convertida a tipo 'category'.
Columna 'Fuente_informacion' convertida a tipo 'category'.
Columna 'País' convertida a tipo 'category'.
Columna 'Banc

In [61]:
# Se detecto que hay una columna (Expectativa_rendimineto) que tiene valores
# basados en un rango de porcentajes y lo que se realizo fue crer dos columna mas y poner valores minimos y maximos del rango
# Crear dos nuevas columnas extrayendo los límites inferior y superior del rango
df_Cambio_Idioma[['Expectativa_rendimiento_min', 'Expectativa_rendimiento_max']] = df_Cambio_Idioma['Expectativa_rendimiento'].str.replace('%', '').str.split('-', expand=True).astype(float)

# Mostrar las nuevas columnas junto con la original para verificar
display(df_Cambio_Idioma[['Expectativa_rendimiento', 'Expectativa_rendimiento_min', 'Expectativa_rendimiento_max']].head())



Expectativa_rendimiento  Expectativa_rendimiento_min  \
0                 20%-30%                        20.00   
1                 30%-40%                        30.00   
2                 10%-20%                        10.00   
3                 10%-20%                        10.00   
4                 10%-20%                        10.00   

   Expectativa_rendimiento_max  
0                        30.00  
1                        40.00  
2                        20.00  
3                        20.00  
4                        20.00

In [62]:
# Verificamos que aparezcan las columnas
df_Cambio_Idioma.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12000 entries, 0 to 11999
Data columns (total 34 columns):
 #   Column                                   Non-Null Count  Dtype         
---  ------                                   --------------  -----         
 0   Genero                                   12000 non-null  category      
 1   Edad                                     12000 non-null  int64         
 2   Vias_de_inversion                        12000 non-null  category      
 3   Fondos_mutuos                            12000 non-null  int64         
 4   Mercado_de_valores                       12000 non-null  int64         
 5   Obligaciones - Bonos Corporativos        12000 non-null  int64         
 6   Bonos_gubernamentales                    12000 non-null  int64         
 7   Depositos_Plazos_fijos                   12000 non-null  int64         
 8   Fondo_de_prevision                       12000 non-null  int64         
 9   Inversión_Oro                          

In [63]:

# Permite instalar la libreria para poder ejecutar el google  translator que no va  a permitir traducir los registros
!pip install googletrans==4.0.0-rc1

In [64]:
# Al Observar que también los registros estan en otro idioma prcedemos a pasarlos al idioma español
# para una mejor compresion de los datos
from googletrans import Translator
import time
import pandas as pd

# Inicializar el traductor
translator = Translator()

# Definir las columnas a excluir de la traducción
columnas_excluir = ['Genero'] # Excluir la columna 'Genero' de la traducción automática

# Obtener las columnas de tipo 'object' a traducir
object_columns = df_Cambio_Idioma.select_dtypes(include='category').columns.tolist()
columnas_a_traducir = [col for col in object_columns if col not in columnas_excluir]

print(f"Columnas a traducir: {columnas_a_traducir}")

# Diccionario para almacenar los mapeos de traducción por columna
mapeos_traduccion = {}

# Traducir valores únicos para cada columna
for columna in columnas_a_traducir:
    valores_unicos = df_Cambio_Idioma[columna].unique()
    print(f"\nTraduciendo valores únicos para la columna: '{columna}'...")

    mapeo_columna = {}
    for valor in valores_unicos:
        # Asegurarse de que el valor no sea nulo y sea una cadena de texto
        if pd.notna(valor) and isinstance(valor, str):
            try:
                # Intentar detectar el idioma de origen y traducir a español
                # source_lang = translator.detect(valor).lang # Opcional: detectar idioma
                translated_text = translator.translate(valor, dest='es').text
                mapeo_columna[valor] = translated_text
                # print(f"  '{valor}' -> '{translated_text}'") # Opcional: imprimir traducciones
                time.sleep(0.1) # Pequeña pausa para evitar sobrecargar el servicio
            except Exception as e:
                print(f"  Error al traducir '{valor}': {e}")
                mapeo_columna[valor] = valor # Mantener el valor original en caso de error
        else:
             mapeo_columna[valor] = valor # Mantener valores no string/nulos


    mapeos_traduccion[columna] = mapeo_columna
    print(f"Traducción de columna '{columna}' completada.")

# Aplicar los mapeos de traducción al DataFrame
print("\nAplicando traducciones al DataFrame...")
for columna, mapeo in mapeos_traduccion.items():
    df_Cambio_Idioma[columna] = df_Cambio_Idioma[columna].map(mapeo)
    print(f"Traducciones aplicadas a la columna '{columna}'.")


print("\n Traducción de columnas de texto completada.")
print("\nVista previa del DataFrame con columnas traducidas:")
display(df_Cambio_Idioma.head())

#  Ruta de salida para guardar el archivo traducido
ruta_salida_traducido = "/content/drive/MyDrive/Finance_Trends_traducido1.csv"

# Guardar el DataFrame traducido como CSV
df_Cambio_Idioma.to_csv(ruta_salida_traducido, index=False, encoding="utf-8-sig")

print(f"\n Archivo traducido guardado correctamente en '{ruta_salida_traducido}'")

Columnas a traducir: ['Vias_de_inversion', 'Part_En_Mercado_de_acciones', 'Factor que influye en Invertir ', 'Objetivo', 'Proposito', 'Duracion', 'Frecuencia_monitoreo', 'Expectativa_rendimiento', 'Tipo_inversion(Corto o largo Plazo,etc)', 'Objetivos_ahorro', 'Razon_equidad', 'Razon_fondos_mutuos', 'Razon_bonos', 'Razon_depositos_fijos', 'Fuente_informacion', 'País', 'Banco', 'Tipo_de_activo', 'Moneda', 'Riesgo']

Traduciendo valores únicos para la columna: 'Vias_de_inversion'...
Traducción de columna 'Vias_de_inversion' completada.

Traduciendo valores únicos para la columna: 'Part_En_Mercado_de_acciones'...
Traducción de columna 'Part_En_Mercado_de_acciones' completada.

Traduciendo valores únicos para la columna: 'Factor que influye en Invertir '...
Traducción de columna 'Factor que influye en Invertir ' completada.

Traduciendo valores únicos para la columna: 'Objetivo'...
Traducción de columna 'Objetivo' completada.

Traduciendo valores únicos para la columna: 'Proposito'...
Tradu

Genero  Edad Vias_de_inversion  Fondos_mutuos  Mercado_de_valores  \
0    Male    29                Sí              2                   4   
1    Male    28                Sí              2                   3   
2  Female    28                Sí              2                   3   
3    Male    19                Sí              2                   3   
4    Male    32                Sí              2                   4   

   Obligaciones - Bonos Corporativos  Bonos_gubernamentales  \
0                                  7                      5   
1                                  6                      5   
2                                  7                      5   
3                                  7                      4   
4                                  7                      5   

   Depositos_Plazos_fijos  Fondo_de_prevision  Inversión_Oro  ...    País  \
0                       3                   1              6  ...  Brasil   
1                       1                   4              7  ...   Chile   
2                       4                   1              6  ...  México   
3                       6                   1              5  ...  España   
4                       3                   1              6  ...  Brasil   

       Banco Tipo_de_activo                Moneda Tasa_interes Riesgo  \
0       Itaú         Fondos  Dólar estadounidense         2.90   Bajo   
1       bbva          Bonos                 euros         3.95   Alto   
2       hsbc  Criptomonedas                   Ars         1.71   Alto   
3  Santander  Criptomonedas                   Ars         5.39   Bajo   
4       hsbc       Acciones                   brl         2.37   Alto   

                          Fecha ID_Registro Expectativa_rendimiento_min  \
0 2022-01-01 00:00:00.000000000           1                       20.00   
1 2022-01-01 02:48:00.840070005           2                       30.00   
2 2022-01-01 05:36:01.680140011           3                       10.00   
3 2022-01-01 08:24:02.520210017           4                       10.00   
4 2022-01-01 11:12:03.360280023           5                       10.00   

  Expectativa_rendimiento_max  
0                       30.00  
1                       40.00  
2                       20.00  
3                       20.00  
4                       20.00  

[5 rows x 34 columns]


 Archivo traducido guardado correctamente en '/content/drive/MyDrive/Finance_Trends_traducido1.csv'


In [65]:
# Mapeo manual para la columna 'Genero'
# El google trslator no pudo cambiar los registros de de la columna genero
# Lo hacemos en forma individual
mapeo_genero = {
    'Male': 'Masculino',
    'Female': 'Femenino'
}

# Aplicar el mapeo a la columna 'Genero'
df_Cambio_Idioma['Genero'] = df_Cambio_Idioma['Genero'].map(mapeo_genero)

# Verificar la traducción
print("Valores únicos en la columna 'Genero' después del mapeo manual:")
print(df_Cambio_Idioma['Genero'].unique())

display(df_Cambio_Idioma.head())

Valores únicos en la columna 'Genero' después del mapeo manual:
['Masculino', 'Femenino']
Categories (2, object): ['Femenino', 'Masculino']


Genero  Edad Vias_de_inversion  Fondos_mutuos  Mercado_de_valores  \
0  Masculino    29                Sí              2                   4   
1  Masculino    28                Sí              2                   3   
2   Femenino    28                Sí              2                   3   
3  Masculino    19                Sí              2                   3   
4  Masculino    32                Sí              2                   4   

   Obligaciones - Bonos Corporativos  Bonos_gubernamentales  \
0                                  7                      5   
1                                  6                      5   
2                                  7                      5   
3                                  7                      4   
4                                  7                      5   

   Depositos_Plazos_fijos  Fondo_de_prevision  Inversión_Oro  ...    País  \
0                       3                   1              6  ...  Brasil   
1                       1                   4              7  ...   Chile   
2                       4                   1              6  ...  México   
3                       6                   1              5  ...  España   
4                       3                   1              6  ...  Brasil   

       Banco Tipo_de_activo                Moneda Tasa_interes Riesgo  \
0       Itaú         Fondos  Dólar estadounidense         2.90   Bajo   
1       bbva          Bonos                 euros         3.95   Alto   
2       hsbc  Criptomonedas                   Ars         1.71   Alto   
3  Santander  Criptomonedas                   Ars         5.39   Bajo   
4       hsbc       Acciones                   brl         2.37   Alto   

                          Fecha ID_Registro Expectativa_rendimiento_min  \
0 2022-01-01 00:00:00.000000000           1                       20.00   
1 2022-01-01 02:48:00.840070005           2                       30.00   
2 2022-01-01 05:36:01.680140011           3                       10.00   
3 2022-01-01 08:24:02.520210017           4                       10.00   
4 2022-01-01 11:12:03.360280023           5                       10.00   

  Expectativa_rendimiento_max  
0                       30.00  
1                       40.00  
2                       20.00  
3                       20.00  
4                       20.00  

[5 rows x 34 columns]

In [66]:
# Le ponemos una breve descripcio a las columnas para
# identificar de que se trata ya que son nombres muy tecnicos
 # Diccionario de descripciones del dataset traducido
descripciones = {
    "Genero": "Identifica el género del participante (por ejemplo, masculino o femenino).",
    "Edad": "Indica la edad del participante en años.",
    "Vias_de_inversion": "Representa los diferentes tipos o canales de inversión utilizados por los encuestados.",
    "Fondos_mutuos": "Muestra si el participante invierte o tiene interés en fondos mutuos.",
    "Mercado_de_valores": "Indica si el participante invierte en el mercado bursátil o de acciones.",
    "Obligaciones - Bonos Corporativos": "Refleja la participación en instrumentos de deuda como obligaciones o bonos corporativos.",
    "Bonos_gubernamentales": "Representa inversiones realizadas en bonos emitidos por el gobierno.",
    "Depositos_Plazos_fijos": "Muestra si el encuestado invierte en depósitos a plazo fijo en instituciones financieras.",
    "Fondo_de_prevision": "Señala si el participante invierte en un fondo de previsión pública o similar.",
    "Inversión_Oro": "Indica si el participante invierte en oro físico o financiero.",
    "Part_En_Mercado_de_acciones": "Evalúa el grado de conocimiento o participación en el mercado de acciones.",
    "Factor que influye en Invertir ": "Describe los factores que influyen en la decisión de inversión del participante.",
    "Objetivo": "Expone el principal objetivo de la inversión (como crecimiento, seguridad o ahorro).",
    "Proposito": "Define el propósito general del ahorro o inversión realizada.",
    "Duracion": "Indica el período de tiempo que el encuestado planea mantener su inversión.",
    "Frecuencia_monitoreo": "Muestra con qué frecuencia el participante revisa o monitorea sus inversiones.",
    "Expectativa_rendimiento": "Refleja las expectativas de rentabilidad o retorno sobre la inversión.",
    "Tipo_inversion(Corto o largo Plazo,etc)": "Clasifica el tipo de inversión preferida (por ejemplo, a corto o largo plazo).",
    "Objetivos_ahorro": "Describe los objetivos de ahorro del participante.",
    "Razon_equidad": "Razones para invertir en acciones.",
    "Razon_fondos_mutuos": "Razones para invertir en fondos mutuos.",
    "Razon_bonos": "Razones para invertir en bonos.",
    "Razon_depositos_fijos": "Razones para invertir en depósitos fijos.",
    "Fuente_informacion": "Identifica la fuente de información utilizada para tomar decisiones de inversión.",
    "País": "País de origen del participante.",
    "Banco": "Banco asociado a las transacciones o inversiones.",
    "Tipo_de_activo": "Categoría del activo financiero.",
    "Moneda": "Moneda en la que se realizan las transacciones.",
    "Tasa_interes": "Tasa de interés asociada a la inversión.",
    "Riesgo": "Nivel de riesgo asociado a la inversión (Bajo, Medio, Alto).",
    "Fecha": "Fecha del registro.",
    "ID_Registro": "Identificador único del registro.",
    "Expectativa_rendimiento_min": "Límite inferior del rango de expectativa de rendimiento.",
    "Expectativa_rendimiento_max": "Límite superior del rango de expectativa de rendimiento."

}
df_info_Descripcion = pd.DataFrame({
    "Columna": df_Cambio_Idioma.columns,
    "Descripción": [descripciones.get(col, "Sin descripción disponible") for col in df_Cambio_Idioma.columns]
})

# Mostrar las primeras filas
print(df_info_Descripcion.head(10))

# Guardar el resumen como CSV en tu Google Drive
ruta_salida = "/content/drive/MyDrive/Finance_Trends_descripciones.csv"
df_info_Descripcion.to_csv(ruta_salida, index=False, encoding="utf-8-sig")

#  Mensaje final
print(" Archivo con descripciones se guardó correctamente en el DataFrame 'df_info_Descripcion'.")

                             Columna  \
0                             Genero   
1                               Edad   
2                  Vias_de_inversion   
3                      Fondos_mutuos   
4                 Mercado_de_valores   
5  Obligaciones - Bonos Corporativos   
6              Bonos_gubernamentales   
7             Depositos_Plazos_fijos   
8                 Fondo_de_prevision   
9                      Inversión_Oro   

                                         Descripción  
0  Identifica el género del participante (por eje...  
1           Indica la edad del participante en años.  
2  Representa los diferentes tipos o canales de i...  
3  Muestra si el participante invierte o tiene in...  
4  Indica si el participante invierte en el merca...  
5  Refleja la participación en instrumentos de de...  
6  Representa inversiones realizadas en bonos emi...  
7  Muestra si el encuestado invierte en depósitos...  
8  Señala si el participante invierte en un fondo...  
9  Indica

In [67]:
# Verificar si hay filas duplicadas
num_duplicados = df_Cambio_Idioma.duplicated().sum()
print(f"Número de filas duplicadas encontradas: {num_duplicados}")

# Si deseas ver las filas duplicadas (opcional)
if num_duplicados > 0:
    print("\nPrimeras 5 filas duplicadas:")
    display(df_Cambio_Idioma[df_Cambio_Idioma.duplicated(keep=False)].head()) # keep=False muestra todos los duplicados

# Para eliminar filas duplicadas (ejecutar solo si deseas eliminarlas)
df_Cambio_Idioma = df_Cambio_Idioma.drop_duplicates()
print(f"\nNúmero de filas después de eliminar duplicados: {len(df_Cambio_Idioma)}")

Número de filas duplicadas encontradas: 0

Número de filas después de eliminar duplicados: 12000


In [68]:
# Verificar si hay valores nulos en cada columna
print("\nConteo de valores nulos por columna:")
display(df_Cambio_Idioma.isnull().sum())


Conteo de valores nulos por columna:


Genero                                     0
Edad                                       0
Vias_de_inversion                          0
Fondos_mutuos                              0
Mercado_de_valores                         0
Obligaciones - Bonos Corporativos          0
Bonos_gubernamentales                      0
Depositos_Plazos_fijos                     0
Fondo_de_prevision                         0
Inversión_Oro                              0
Part_En_Mercado_de_acciones                0
Factor que influye en Invertir             0
Objetivo                                   0
Proposito                                  0
Duracion                                   0
Frecuencia_monitoreo                       0
Expectativa_rendimiento                    0
Tipo_inversion(Corto o largo Plazo,etc)    0
Objetivos_ahorro                           0
Razon_equidad                              0
Razon_fondos_mutuos                        0
Razon_bonos                                0
Razon_depositos_fijos                      0
Fuente_informacion                         0
País                                       0
Banco                                      0
Tipo_de_activo                             0
Moneda                                     0
Tasa_interes                               0
Riesgo                                     0
Fecha                                      0
ID_Registro                                0
Expectativa_rendimiento_min                0
Expectativa_rendimiento_max                0
dtype: int64

### Creación de Nuevas Variables

In [69]:
# 1. Crear Rango de Edad
# Definir los rangos de edad
bins = [18, 25, 35, 100] # Ejemplo: 18-24, 25-34, 35-99 o hasta <100
labels = ['18-24', '25-34', '35+'] # Añadimos una etiqueta para el último rango

# Crear la nueva columna 'Rango_Edad'
df_Cambio_Idioma['Rango_Edad'] = pd.cut(df_Cambio_Idioma['Edad'], bins=bins, labels=labels, right=False, include_lowest=True)

print("Distribución de la nueva columna 'Rango_Edad':")
display(df_Cambio_Idioma['Rango_Edad'].value_counts())

Distribución de la nueva columna 'Rango_Edad':


Rango_Edad
25-34    8737
18-24    2622
35+       641
Name: count, dtype: int64

In [70]:
# 2. Extraer Año y Mes de la Fecha
df_Cambio_Idioma['Año_Registro'] = df_Cambio_Idioma['Fecha'].dt.year
df_Cambio_Idioma['Mes_Registro'] = df_Cambio_Idioma['Fecha'].dt.month_name() # Quitamos locale='es'

print("\nDistribución de la nueva columna 'Año_Registro':")
display(df_Cambio_Idioma['Año_Registro'].value_counts().sort_index())

print("\nDistribución de la nueva columna 'Mes_Registro':")
display(df_Cambio_Idioma['Mes_Registro'].value_counts())


Distribución de la nueva columna 'Año_Registro':


Año_Registro
2022    3129
2023    3128
2024    3137
2025    2606
Name: count, dtype: int64


Distribución de la nueva columna 'Mes_Registro':


Mes_Registro
January      1064
May          1064
July         1063
March        1063
August       1062
October      1062
September    1029
April        1028
June         1028
February      967
December      798
November      772
Name: count, dtype: int64

In [71]:
# Mapeo manual para traducir los nombres de los meses a español
mapeo_meses_espanol = {
    'January': 'Enero',
    'February': 'Febrero',
    'March': 'Marzo',
    'April': 'Abril',
    'May': 'Mayo',
    'June': 'Junio',
    'July': 'Julio',
    'August': 'Agosto',
    'September': 'Septiembre',
    'October': 'Octubre',
    'November': 'Noviembre',
    'December': 'Diciembre'
}

# Aplicar el mapeo a la columna 'Mes_Registro'
df_Cambio_Idioma['Mes_Registro'] = df_Cambio_Idioma['Mes_Registro'].map(mapeo_meses_espanol)

# Verificar la traducción mostrando los valores únicos
print("Valores únicos en la columna 'Mes_Registro' después del mapeo manual:")
print(df_Cambio_Idioma['Mes_Registro'].unique())

display(df_Cambio_Idioma.head())

Valores únicos en la columna 'Mes_Registro' después del mapeo manual:
['Enero' 'Febrero' 'Marzo' 'Abril' 'Mayo' 'Junio' 'Julio' 'Agosto'
 'Septiembre' 'Octubre' 'Noviembre' 'Diciembre']


Genero  Edad Vias_de_inversion  Fondos_mutuos  Mercado_de_valores  \
0  Masculino    29                Sí              2                   4   
1  Masculino    28                Sí              2                   3   
2   Femenino    28                Sí              2                   3   
3  Masculino    19                Sí              2                   3   
4  Masculino    32                Sí              2                   4   

   Obligaciones - Bonos Corporativos  Bonos_gubernamentales  \
0                                  7                      5   
1                                  6                      5   
2                                  7                      5   
3                                  7                      4   
4                                  7                      5   

   Depositos_Plazos_fijos  Fondo_de_prevision  Inversión_Oro  ...  \
0                       3                   1              6  ...   
1                       1                   4              7  ...   
2                       4                   1              6  ...   
3                       6                   1              5  ...   
4                       3                   1              6  ...   

                 Moneda Tasa_interes Riesgo                         Fecha  \
0  Dólar estadounidense         2.90   Bajo 2022-01-01 00:00:00.000000000   
1                 euros         3.95   Alto 2022-01-01 02:48:00.840070005   
2                   Ars         1.71   Alto 2022-01-01 05:36:01.680140011   
3                   Ars         5.39   Bajo 2022-01-01 08:24:02.520210017   
4                   brl         2.37   Alto 2022-01-01 11:12:03.360280023   

  ID_Registro Expectativa_rendimiento_min Expectativa_rendimiento_max  \
0           1                       20.00                       30.00   
1           2                       30.00                       40.00   
2           3                       10.00                       20.00   
3           4                       10.00                       20.00   
4           5                       10.00                       20.00   

  Rango_Edad Año_Registro Mes_Registro  
0      25-34         2022        Enero  
1      25-34         2022        Enero  
2      25-34         2022        Enero  
3      18-24         2022        Enero  
4      25-34         2022        Enero  

[5 rows x 37 columns]

## 3. Carga de los Datos , etapa final del ETL (LOAD)

In [72]:
# Ruta de salida para guardar el archivo limpio y transformado
ruta_salida_limpio = "/content/drive/MyDrive/Finance_Trends_limpio_transformado1.csv"

# Guardar el DataFrame limpio y transformado como CSV
df_Cambio_Idioma.to_csv(ruta_salida_limpio, index=False, encoding="utf-8-sig")

print(f"Archivo limpio y transformado guardado correctamente en '{ruta_salida_limpio}'")

Archivo limpio y transformado guardado correctamente en '/content/drive/MyDrive/Finance_Trends_limpio_transformado1.csv'


#EDA (Analisis Exploratorio de Datos)

In [73]:


# count: Muestra cuántos valores no nulos hay en cada columna.
# mean: Es el promedio de los valores en la columna. Por ejemplo, la edad promedio de los participantes es aproximadamente 27.78 años.
# std: Es la desviación estándar. Mide cuánto se dispersan los valores de la columna respecto al promedio. Una desviación estándar alta indica que los valores están más dispersos.
# min: Es el valor mínimo que se encuentra en la columna. Por ejemplo, la edad mínima es 18 años.
# 25%: Es el primer cuartil (Q1). Indica que el 25% de los datos son menores o iguales a este valor.
# 50%: Es la mediana o segundo cuartil (Q2). Es el valor central de los datos cuando están ordenados. El 50% de los datos son menores o iguales a este valor (también es el mismo que el mean si los datos tienen una distribución perfectamente simétrica).
# 75%: Es el tercer cuartil (Q3). Indica que el 75% de los datos son menores o iguales a este valor.
# max: Es el valor máximo que se encuentra en la columna. Por ejemplo, la edad máxima es 38 años.

#Interpretación:
# Si las tasas de interés o rendimientos tienen mucha variación.
# Si hay (valores atípicos).
# Cómo se distribuyen los datos en columnas numéricas.


df_Cambio_Idioma.describe()

Edad  Fondos_mutuos  Mercado_de_valores  \
count 12000.00       12000.00            12000.00   
mean     27.78           2.53                3.46   
min      18.00           1.00                1.00   
25%      25.00           2.00                3.00   
50%      28.00           2.00                4.00   
75%      31.00           3.00                4.00   
max      38.00           7.00                6.00   
std       4.06           1.17                1.11   

       Obligaciones - Bonos Corporativos  Bonos_gubernamentales  \
count                           12000.00               12000.00   
mean                                5.78                   4.67   
min                                 1.00                   1.00   
25%                                 5.00                   4.00   
50%                                 7.00                   5.00   
75%                                 7.00                   5.00   
max                                 7.00                   7.00   
std                                 1.64                   1.33   

       Depositos_Plazos_fijos  Fondo_de_prevision  Inversión_Oro  \
count                12000.00            12000.00       12000.00   
mean                     3.56                2.02           5.98   
min                      1.00                1.00           2.00   
25%                      2.00                1.00           6.00   
50%                      3.00                1.00           6.00   
75%                      5.00                3.00           7.00   
max                      7.00                6.00           7.00   
std                      1.76                1.59           1.12   

       Tasa_interes                          Fecha  ID_Registro  \
count      12000.00                          12000     12000.00   
mean           6.30  2023-12-02 00:00:00.000000256      6000.50   
min            0.50            2022-01-01 00:00:00         1.00   
25%            3.44            2022-12-17 00:00:00      3000.75   
50%            6.32            2023-12-02 00:00:00      6000.50   
75%            9.19            2024-11-16 00:00:00      9000.25   
max           12.00            2025-11-01 00:00:00     12000.00   
std            3.33                            NaN      3464.25   

       Expectativa_rendimiento_min  Expectativa_rendimiento_max  Año_Registro  
count                     12000.00                     12000.00      12000.00  
mean                         20.13                        30.13       2023.43  
min                          10.00                        20.00       2022.00  
25%                          10.00                        20.00       2022.00  
50%                          20.00                        30.00       2023.00  
75%                          30.00                        40.00       2024.00  
max                          30.00                        40.00       2025.00  
std                           8.21                         8.21          1.10

In [74]:
# Interpretación:
# El método .info() permite ver cuántas filas y columnas tiene el dataset, los tipos de datos (numéricos, de texto o fecha) y si hay valores faltantes.
# Esto ayuda a confirmar que los tipos se transformaron correctamente en la etapa ETL.
df_Cambio_Idioma.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12000 entries, 0 to 11999
Data columns (total 37 columns):
 #   Column                                   Non-Null Count  Dtype         
---  ------                                   --------------  -----         
 0   Genero                                   12000 non-null  category      
 1   Edad                                     12000 non-null  int64         
 2   Vias_de_inversion                        12000 non-null  category      
 3   Fondos_mutuos                            12000 non-null  int64         
 4   Mercado_de_valores                       12000 non-null  int64         
 5   Obligaciones - Bonos Corporativos        12000 non-null  int64         
 6   Bonos_gubernamentales                    12000 non-null  int64         
 7   Depositos_Plazos_fijos                   12000 non-null  int64         
 8   Fondo_de_prevision                       12000 non-null  int64         
 9   Inversión_Oro                          

In [75]:
# Interpretacion
# Permite ver las cabeceras de todas las colummas y si realmente se aplicaron todos los cambios y transformaciones que reaizamos
df_Cambio_Idioma.head()

Genero  Edad Vias_de_inversion  Fondos_mutuos  Mercado_de_valores  \
0  Masculino    29                Sí              2                   4   
1  Masculino    28                Sí              2                   3   
2   Femenino    28                Sí              2                   3   
3  Masculino    19                Sí              2                   3   
4  Masculino    32                Sí              2                   4   

   Obligaciones - Bonos Corporativos  Bonos_gubernamentales  \
0                                  7                      5   
1                                  6                      5   
2                                  7                      5   
3                                  7                      4   
4                                  7                      5   

   Depositos_Plazos_fijos  Fondo_de_prevision  Inversión_Oro  ...  \
0                       3                   1              6  ...   
1                       1                   4              7  ...   
2                       4                   1              6  ...   
3                       6                   1              5  ...   
4                       3                   1              6  ...   

                 Moneda Tasa_interes Riesgo                         Fecha  \
0  Dólar estadounidense         2.90   Bajo 2022-01-01 00:00:00.000000000   
1                 euros         3.95   Alto 2022-01-01 02:48:00.840070005   
2                   Ars         1.71   Alto 2022-01-01 05:36:01.680140011   
3                   Ars         5.39   Bajo 2022-01-01 08:24:02.520210017   
4                   brl         2.37   Alto 2022-01-01 11:12:03.360280023   

  ID_Registro Expectativa_rendimiento_min Expectativa_rendimiento_max  \
0           1                       20.00                       30.00   
1           2                       30.00                       40.00   
2           3                       10.00                       20.00   
3           4                       10.00                       20.00   
4           5                       10.00                       20.00   

  Rango_Edad Año_Registro Mes_Registro  
0      25-34         2022        Enero  
1      25-34         2022        Enero  
2      25-34         2022        Enero  
3      18-24         2022        Enero  
4      25-34         2022        Enero  

[5 rows x 37 columns]

In [76]:
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd

# Calcular la tasa promedio de interés por país y ordenarla
tasa_promedio_por_pais = df_Cambio_Idioma.groupby("País")["Tasa_interes"].mean().sort_values(ascending=False)

# Mostrar los resultados en formato de tabla (Serie de pandas)
print("Tasa promedio de interés por país:")
display(tasa_promedio_por_pais)

# Intepretacion
# El análisis muestra que, según el indicador evaluado, los países presentan valores promedio bastante similares,
# con una ligera ventaja para Estados Unidos.
# Esto sugiere que no existen grandes diferencias en el comportamiento financiero general del conjunto analizado,
#  aunque ciertos países como Colombia presentan un rendimiento algo inferior.”
# esto permite comparar políticas financieras o entornos económicos distintos.
# Los países con tasas más altas pueden tener inflación elevada o mayor riesgo crediticio.

Tasa promedio de interés por país:


/tmp/ipython-input-2022832319.py:6: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  tasa_promedio_por_pais = df_Cambio_Idioma.groupby("País")["Tasa_interes"].mean().sort_values(ascending=False)


País
Eeuu        6.42
Argentina   6.37
Brasil      6.33
México      6.30
Chile       6.25
España      6.24
Colombia    6.18
Name: Tasa_interes, dtype: float64

In [77]:
# Interpretación

# La tabla nos muestra que, en promedio, las Criptomonedas tuvieron la tasa de interés promedio más baja (aproximadamente 6.22%).
# Las Materias primas y las Acciones tuvieron tasas de interés promedio un poco más altas, pero aún por debajo del promedio general.
# Los Fondos tuvieron una tasa promedio ligeramente más alta que las acciones.
# Los Bonos tuvieron la tasa de interés promedio más alta (aproximadamente 6.41%) entre todos los tipos de activo listados.
# En resumen, según estos datos, si solo nos  fijamos en la tasa de interés promedio, los Bonos fueron el tipo de activo
# con el retorno promedio más alto, mientras que las Criptomonedas tuvieron el más bajo.
# Esta información es útil para entender el rendimiento histórico promedio de diferentes tipos de inversión

import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd

# Calcular la tasa promedio de interés por tipo de activo y ordenamos
tasa_promedio_por_activo = df_Cambio_Idioma.groupby("Tipo_de_activo")["Tasa_interes"].mean().sort_values()

# Mostrar los resultados en formato de tabla
print("Tasa promedio por tipo de activo:")
display(tasa_promedio_por_activo)



Tasa promedio por tipo de activo:


/tmp/ipython-input-2680202259.py:16: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  tasa_promedio_por_activo = df_Cambio_Idioma.groupby("Tipo_de_activo")["Tasa_interes"].mean().sort_values()


Tipo_de_activo
Criptomonedas     6.22
Materias primas   6.26
Acciones          6.29
Fondos            6.31
Bonos             6.41
Name: Tasa_interes, dtype: float64

In [78]:
# Interpretacion
# Permite ver la evolución mensual de la Tasa de Interés promedio.
# Esto ayuda a detectar tendencias estacionales o cambios significativos en el tiempo.
# Un aumento podría indicar condiciones económicas cambiantes o políticas monetarias.
# Una disminución podría sugerir lo contrario.
# La tabla revela  si la tasa ha sido volátil o estable.

import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd

# Asegurarse de que la columna 'Fecha' es de tipo datetime
df_Cambio_Idioma["Fecha"] = pd.to_datetime(df_Cambio_Idioma["Fecha"])
# Opcional: ordenar por fecha si es necesario para el análisis de series de tiempo
df_Cambio_Idioma.sort_values("Fecha", inplace=True)

# Calcular la tasa promedio de interés agrupada por mes
# Convertimos el resultado a DataFrame para mostrarlo mejor si es necesario
tasa_promedio_mensual = df_Cambio_Idioma.groupby(df_Cambio_Idioma["Fecha"].dt.to_period("M"))["Tasa_interes"].mean()

print("Evolución mensual de la Tasa de Interés promedio:")
display(tasa_promedio_mensual)



Evolución mensual de la Tasa de Interés promedio:


Fecha
2022-01   6.25
2022-02   5.90
2022-03   6.37
2022-04   6.14
2022-05   6.13
2022-06   6.36
2022-07   6.29
2022-08   6.04
2022-09   6.41
2022-10   6.18
2022-11   6.38
2022-12   6.55
2023-01   6.26
2023-02   6.46
2023-03   6.54
2023-04   6.57
2023-05   6.22
2023-06   6.29
2023-07   6.29
2023-08   6.56
2023-09   6.18
2023-10   5.90
2023-11   6.38
2023-12   6.24
2024-01   6.16
2024-02   5.88
2024-03   6.54
2024-04   6.35
2024-05   6.86
2024-06   6.18
2024-07   6.39
2024-08   5.98
2024-09   6.37
2024-10   6.16
2024-11   6.16
2024-12   6.30
2025-01   6.39
2025-02   6.27
2025-03   6.20
2025-04   6.08
2025-05   7.07
2025-06   6.59
2025-07   5.99
2025-08   6.51
2025-09   6.06
2025-10   6.29
2025-11   4.77
Freq: M, Name: Tasa_interes, dtype: float64

In [79]:
# Interpretacion
# Esta tabla permite ver el promedio de como invierten la personas en diferentes activos separados por su pai y rango de edad
# Al ver esto podemos compaarar :
# ENTRE PAISES : ¿ Invierte la gente del mismo rango de edad ce manera similar en diferentes paises ?
# ENTRE EDADES : ¿ Dentro de un mismo pai  cambian las preferencias de inversion promedio a medida que la
# gente es mas joven o de mayor edad?

# Agrupación 1: Por País y Rango de Edad
grouping_cols_1 = ['País', 'Rango_Edad']
cols_to_aggregate = df_Cambio_Idioma.columns[df_Cambio_Idioma.columns.get_loc('Fondos_mutuos') : df_Cambio_Idioma.columns.get_loc('Inversión_Oro') + 1].tolist()

print(f"Agrupación 1: Columnas para agrupar: {grouping_cols_1}")
print(f"Columnas a agregar: {cols_to_aggregate}")

# Realizar la agrupación y aplicar solo la agregación 'mean'
grouped_data_1 = df_Cambio_Idioma.groupby(grouping_cols_1, observed=False)[cols_to_aggregate].agg('mean')

# Opcional: Aplanar el índice de columnas y renombrar para mayor claridad
grouped_data_1.columns = grouped_data_1.columns.get_level_values(0)

# Formatear los números a 2 decimales para una mejor visualización
pd.options.display.float_format = '{:.2f}'.format


print("\nResultado de la Agrupación 1 (Promedio por País y Rango de Edad):")
display(grouped_data_1.head(40))

Agrupación 1: Columnas para agrupar: ['País', 'Rango_Edad']
Columnas a agregar: ['Fondos_mutuos', 'Mercado_de_valores', 'Obligaciones - Bonos Corporativos', 'Bonos_gubernamentales', 'Depositos_Plazos_fijos', 'Fondo_de_prevision', 'Inversión_Oro']

Resultado de la Agrupación 1 (Promedio por País y Rango de Edad):


Fondos_mutuos  Mercado_de_valores  \
País      Rango_Edad                                      
Argentina 18-24                2.60                3.06   
          25-34                2.51                3.60   
          35+                  2.31                3.49   
Brasil    18-24                2.68                2.97   
          25-34                2.53                3.58   
          35+                  2.66                3.69   
Chile     18-24                2.66                3.07   
          25-34                2.52                3.57   
          35+                  2.33                3.65   
Colombia  18-24                2.54                2.92   
          25-34                2.50                3.60   
          35+                  2.55                3.62   
Eeuu      18-24                2.84                3.06   
          25-34                2.47                3.62   
          35+                  2.26                3.65   
España    18-24                2.48                2.83   
          25-34                2.52                3.58   
          35+                  2.43                3.58   
México    18-24                2.61                3.05   
          25-34                2.52                3.55   
          35+                  2.74                3.72   

                      Obligaciones - Bonos Corporativos  \
País      Rango_Edad                                      
Argentina 18-24                                    4.94   
          25-34                                    6.05   
          35+                                      5.85   
Brasil    18-24                                    4.82   
          25-34                                    6.08   
          35+                                      5.73   
Chile     18-24                                    4.75   
          25-34                                    6.02   
          35+                                      6.04   
Colombia  18-24                                    4.90   
          25-34                                    6.06   
          35+                                      5.68   
Eeuu      18-24                                    4.65   
          25-34                                    6.04   
          35+                                      6.19   
España    18-24                                    4.67   
          25-34                                    6.09   
          35+                                      5.81   
México    18-24                                    4.86   
          25-34                                    6.12   
          35+                                      5.72   

                      Bonos_gubernamentales  Depositos_Plazos_fijos  \
País      Rango_Edad                                                  
Argentina 18-24                        4.67                    3.89   
          25-34                        4.68                    3.34   
          35+                          4.01                    4.49   
Brasil    18-24                        4.69                    3.99   
          25-34                        4.69                    3.37   
          35+                          3.97                    4.40   
Chile     18-24                        4.70                    3.97   
          25-34                        4.71                    3.38   
          35+                          4.19                    4.22   
Colombia  18-24                        4.79                    4.10   
          25-34                        4.67                    3.42   
          35+                          3.92                    4.62   
Eeuu      18-24                        4.71                    4.02   
          25-34                        4.67                    3.36   
          35+                          4.30                    3.98   
España    18-24                        4.87                    4.20   
          25-34             

#PREGUNTAS DE NEGOCIO

### Pregunta de Negocio 1: ¿En qué tipo de activo invierten más (en promedio, según las puntuaciones) las personas de cada rango de edad?

In [80]:
# Para responder a esta pregunta, agruparemos por Rango de Edad
# y calcularemos el promedio de las columnas de inversión.

grouping_col = 'Rango_Edad'
cols_to_aggregate = df_Cambio_Idioma.columns[df_Cambio_Idioma.columns.get_loc('Fondos_mutuos') : df_Cambio_Idioma.columns.get_loc('Inversión_Oro') + 1].tolist()

print(f"Agrupando por: {grouping_col}")
print(f"Columnas de inversión a promediar: {cols_to_aggregate}")

# Realizar la agrupación por Rango_Edad y calcular el promedio
average_investment_by_age = df_Cambio_Idioma.groupby(grouping_col, observed=False)[cols_to_aggregate].agg('mean')

# Mostrar la tabla con los promedios
print("\nTabla de Evidencia: Promedio de inversión por Rango de Edad")
display(average_investment_by_age)

Agrupando por: Rango_Edad
Columnas de inversión a promediar: ['Fondos_mutuos', 'Mercado_de_valores', 'Obligaciones - Bonos Corporativos', 'Bonos_gubernamentales', 'Depositos_Plazos_fijos', 'Fondo_de_prevision', 'Inversión_Oro']

Tabla de Evidencia: Promedio de inversión por Rango de Edad


Fondos_mutuos  Mercado_de_valores  \
Rango_Edad                                      
18-24                2.63                3.00   
25-34                2.51                3.59   
35+                  2.47                3.63   

            Obligaciones - Bonos Corporativos  Bonos_gubernamentales  \
Rango_Edad                                                             
18-24                                    4.80                   4.73   
25-34                                    6.07                   4.70   
35+                                      5.86                   4.05   

            Depositos_Plazos_fijos  Fondo_de_prevision  Inversión_Oro  
Rango_Edad                                                             
18-24                         4.02                2.90           5.92  
25-34                         3.36                1.76           6.02  
35+                           4.35                2.02           5.63

**Interpretación de la Pregunta de Negocio 1:**

Esta tabla te muestra, para cada grupo de edad (18-24, 25-34, 35+), cuál es el promedio de las puntuaciones en cada tipo de inversión (Fondos Mutuos, Mercado de Valores, etc.).

*   Puedes ver qué tipo de inversión tiene el promedio más alto dentro de cada rango de edad.
*   Puedes comparar los promedios de un mismo tipo de inversión (por ejemplo, "Inversión_Oro") a través de los diferentes rangos de edad para ver si la preferencia cambia con la edad.

Por ejemplo, si ves que la puntuación promedio para "Mercado_de_valores" es más alta en el rango "25-34" que en "18-24" o "35+", podrías concluir que las personas de 25 a 34 años muestran, en promedio, un mayor interés o participación en el mercado de valores

Análisis Adicional de la Pregunta de Negocio 1: Inversiones con Promedios Más Altos y Bajos por Rango de Edad**

Basándonos en la tabla de promedio de inversión por Rango de Edad , podemos identificar las inversiones con las puntuaciones promedio más altas y más bajas dentro de cada grupo de edad:

*   **Rango de Edad 18-24:**
    *   La inversión con el promedio más alto es **Inversión_Oro** (Promedio: 5.92).
    *   La inversión con el promedio más bajo es **Fondo_de_prevision** (Promedio: 2.90).

*   **Rango de Edad 25-34:**
    *   La inversión con el promedio más alto es **Obligaciones - Bonos Corporativos** (Promedio: 6.07).
    *   La inversión con el promedio más bajo es **Fondo_de_prevision** (Promedio: 1.76).

*   **Rango de Edad 35+:**
    *   La inversión con el promedio más alto es **Obligaciones - Bonos Corporativos** (Promedio: 5.86).
    *   La inversión con el promedio más bajo es **Fondo_de_prevision** (Promedio: 2.02).

**Conclusiones Clave:**

*   **Fondo_de_prevision** consistentemente tenemos el promedio de puntuación más bajo en todos los rangos de edad en este dataset.
*   Las **Obligaciones - Bonos Corporativos** tienen los promedios más altos en los rangos de edad de 25-34 y 35+.
*   La **Inversión_Oro** es la más alta en el rango de edad más joven (18-24).

Esto sugiere que el interés o la participación en el Fondo de Previsión es generalmente bajo, mientras que las Obligaciones/Bonos Corporativos y la Inversión en Oro son relativamente más populares, variando la preferencia principal según el rango de edad.

### Pregunta de Negocio 2: ¿Hay una diferencia clara en la tasa de interés promedio por nivel de riesgo Bajo, Medio y Alto y tipo de activo?


Para complementar la estadística general de tasa de interés por nivel de riesgo, podemos observar la tasa promedio para cada tipo de activo dentro de cada categoría de riesgo. Esto nos da una visión más detallada de cómo las tasas varían para inversiones específicas.

In [81]:
# Agrupar por Riesgo y Tipo_de_activo y calcular el promedio de Tasa_interes
risk_asset_interest_mean = df_Cambio_Idioma.groupby(['Riesgo', 'Tipo_de_activo'], observed=False)['Tasa_interes'].mean()

# Opcional: Desapilar la tabla para que sea más fácil de leer, con Riesgo como filas y Tipo_de_activo como columnas
risk_asset_interest_pivot = risk_asset_interest_mean.unstack()

# Reordenar las filas por nivel de riesgo (Bajo, Medio, Alto)
risk_asset_interest_pivot = risk_asset_interest_pivot.loc[['Bajo', 'Medio', 'Alto']]


print("Tasa de Interés Promedio por Nivel de Riesgo y Tipo de Activo:")
display(risk_asset_interest_pivot)

Tasa de Interés Promedio por Nivel de Riesgo y Tipo de Activo:


Tipo_de_activo  Acciones  Bonos  Materias primas  Criptomonedas  Fondos
Riesgo                                                                 
Bajo                6.15   6.39             6.34           6.31    6.30
Medio               6.35   6.43             6.26           6.09    6.24
Alto                6.38   6.40             6.17           6.25    6.38

**Interpretación de la Tasa de Interés Promedio por Nivel de Riesgo y Tipo de Activo:**

Esta tabla  muestra la tasa de interés promedio para cada combinación de Nivel de Riesgo (filas) y Tipo de Activo (columnas).

*   **Comparar por Tipo de Activo dentro de un Riesgo:** Puedes elegir un nivel de riesgo (por ejemplo, 'Alto') y comparar las tasas promedio de los diferentes tipos de activo en esa fila (Criptomonedas, Bonos, etc.). Esto te dirá qué tipo de activo, dentro de ese nivel de riesgo, tuvo la tasa promedio más alta o más baja.
*   **Comparar por Riesgo dentro de un Tipo de Activo:** Puedes elegir un tipo de activo (por ejemplo, 'Bonos') y comparar sus tasas promedio a través de los diferentes niveles de riesgo (Bajo, Medio, Alto) en esa columna. Esto te mostrará si el mismo tipo de activo ofrece tasas promedio diferentes dependiendo de cómo se clasifique su riesgo.

Observando la tabla, puedes identificar patrones más específicos que la estadística general por riesgo. Por ejemplo, se podrías notar si un tipo de activo en particular tiene una tasa promedio significativamente más alta en el nivel de riesgo 'Alto' en comparación con otros tipos de activo en ese mismo nivel.

**Interpretación de la Pregunta de Negocio 2:**

Esta tabla muestra estadísticas clave (Promedio, Mediana, Mínimo, Máximo y Desviación Estándar) de la Tasa de Interés para cada nivel de Riesgo (Bajo, Medio, Alto).

Al observar la tabla:

*   **Promedio y Mediana:** Compara los valores en las columnas 'mean' (promedio) y 'median' (mediana) para cada nivel de riesgo. Si estos valores aumentan a medida que pasas de 'Bajo' a 'Medio' y a 'Alto', sugiere que, en promedio, las inversiones de mayor riesgo tienden a ofrecer tasas de interés más altas.
*   **Mínimo y Máximo:** Las columnas 'min' y 'max' te muestran el rango completo de tasas de interés encontradas dentro de cada nivel de riesgo.
*   **Desviación Estándar:** La columna 'std' te indica cuán dispersas están las tasas de interés dentro de cada grupo de riesgo. Una desviación estándar más alta significa que las tasas varían más en ese grupo.

En resumen, esta tabla te permite ver numéricamente si existe una relación entre el nivel de riesgo y las tasas de interés, y cómo se distribuyen esas tasas en cada categoría de riesgo.

### Pregunta de Negocio 3: ¿Cuáles son los tipos de activo más populares (los que tienen el promedio de puntuación más alto) en cada país?

In [82]:
# Para responder a esta pregunta, agruparemos por País
# y calcularemos el promedio de las columnas de inversión.

grouping_col = 'País'
cols_to_aggregate = df_Cambio_Idioma.columns[df_Cambio_Idioma.columns.get_loc('Fondos_mutuos') : df_Cambio_Idioma.columns.get_loc('Inversión_Oro') + 1].tolist()

print(f"Agrupando por: {grouping_col}")
print(f"Columnas de inversión a promediar: {cols_to_aggregate}")

# Realizar la agrupación por País y calcular el promedio
average_investment_by_country = df_Cambio_Idioma.groupby(grouping_col, observed=False)[cols_to_aggregate].agg('mean')

# Mostrar la tabla con los promedios
print("\nTabla de Evidencia: Promedio de inversión por País")
display(average_investment_by_country)

Agrupando por: País
Columnas de inversión a promediar: ['Fondos_mutuos', 'Mercado_de_valores', 'Obligaciones - Bonos Corporativos', 'Bonos_gubernamentales', 'Depositos_Plazos_fijos', 'Fondo_de_prevision', 'Inversión_Oro']

Tabla de Evidencia: Promedio de inversión por País


Fondos_mutuos  Mercado_de_valores  \
País                                           
Argentina           2.52                3.48   
Brasil              2.57                3.45   
Chile               2.54                3.47   
Colombia            2.51                3.45   
Eeuu                2.54                3.50   
España              2.51                3.42   
México              2.55                3.45   

           Obligaciones - Bonos Corporativos  Bonos_gubernamentales  \
País                                                                  
Argentina                               5.79                   4.64   
Brasil                                  5.78                   4.65   
Chile                                   5.74                   4.68   
Colombia                                5.79                   4.65   
Eeuu                                    5.76                   4.65   
España                                  5.77                   4.73   
México                                  5.81                   4.68   

           Depositos_Plazos_fijos  Fondo_de_prevision  Inversión_Oro  
País                                                                  
Argentina                    3.52                2.06           5.99  
Brasil                       3.56                2.02           5.97  
Chile                        3.55                2.02           6.01  
Colombia                     3.63                2.01           5.96  
Eeuu                         3.53                2.03           5.99  
España                       3.59                2.04           5.94  
México                       3.52                2.00           6.00

**Interpretación de la Pregunta de Negocio 3:**

Esta tabla te muestra, para cada país, cuál es el promedio de las puntuaciones en cada tipo de inversión (Fondos Mutuos, Mercado de Valores, etc.).

*   Para cada país (cada fila), puedes ver qué tipo de inversión tiene el promedio más alto. Ese sería el "más popular" o con mayor puntuación promedio en ese país según este dataset.
*   Puedes comparar los promedios de un mismo tipo de inversión a través de los diferentes países para ver dónde es más o menos popular.

Por ejemplo, si en la fila de "Brasil" la columna "Inversión_Oro" tiene el promedio más alto comparado con las otras columnas en esa fila, podrías decir que la Inversión en Oro es la más popular en Brasil según estos datos.